packages/timeline-state-resolver/src/integrations/tcpSend/tcpConnection.ts

```diff
@@ -0,0 +1,154 @@
+	 * Is set when the connection is active.
+	 * is set to undefined if disconnect() has been called (then do not try to reconnect)
+	 */
+	private activeOptions: TCPSendOptions | undefined = undefined
+	private _tcpClient: Socket | null = null
+
+	private _connected = false
+	private _retryConnectTimeout: NodeJS.Timer
+
+	get connected(): boolean {
+		return this._connected
+	}
+	activate(options: TCPSendOptions): void {
+		this.activeOptions = options
+
+		this.ensureConnection().catch((err) => {
+			this.emit('error', 'activate TCP', err)
+		})
+	}
+	async ensureConnection(): Promise<Socket> {
+		if (!this.activeOptions) throw new Error('TCP connection not activated')
+		const activeOptions = this.activeOptions
+
+		return new Promise((resolve, reject) => {
+			if (!this._tcpClient) {
+				this._tcpClient = new Socket()
+				this._tcpClient.on('connect', () => {
+					this._setConnected(true)
+				})
+				this._tcpClient.on('close', () => {
+					this._setConnected(false)
+				})
+				this._tcpClient.on('end', () => {
+					this._setConnected(false)
+				})
+			}
+			const tcpClient: Socket = this._tcpClient
+
+			if (!this.connected) {
+				tcpClient.connect(activeOptions.port, activeOptions.host, () => {
+					resolve(tcpClient)
+				})
+				setTimeout(() => {
+					reject(new Error(`TCP timeout: Unable to connect to ${activeOptions.host}:${activeOptions.port}`))
+				}, TIMEOUT)
+			} else {
+				resolve(tcpClient)
+			}
+		})
+	}
+	async deactivate(): Promise<void> {
+		this.activeOptions = undefined // prevent reconnecting
+
+		if (this._tcpClient) {
+			const tcpClient = this._tcpClient
+
+			await new Promise<void>((resolve) => {
+				tcpClient.once('close', () => {
+					resolve()
+				})
+				tcpClient.once('end', () => {
+					resolve()
+				})
+
+				tcpClient.end()
+				setTimeout(() => {
+					resolve()
+				}, TIMEOUT)
+				setTimeout(() => {
+					if (this._tcpClient) {
+						// Forcefully destroy the connection:
+						this._tcpClient.destroy()
+					}
+				}, Math.floor(TIMEOUT / 2))
+			})
+		}
+
+		if (this._tcpClient) {
+			this._tcpClient.removeAllListeners('connect')
+			this._tcpClient.removeAllListeners('close')
+			this._tcpClient.removeAllListeners('end')
+			this._tcpClient.removeAllListeners('error')
+
+			this._tcpClient = null
+		}
+		this._setConnected(false)
+	}
+	async reconnect() {
+		if (!this.activeOptions) throw new Error('TCP connection not activated')
+
+		const options = this.activeOptions
+		await this.deactivate()
+		this.activate(options)
+	}
+	async sendTCPMessage(message: string): Promise<void> {
+		if (!this.activeOptions) throw new Error('TCP connection not activated')
+		const tcpClient = await this.ensureConnection()
+
+		tcpClient.write(Buffer.from(message, this.activeOptions.bufferEncoding))
+	}
+
+	private _setConnected(connected: boolean) {
+		if (this._connected !== connected) {
+			this._connected = connected
+			this._connectionChanged()
+
+			if (!connected) {
+				this._tcpClient = null
```

Comment:
It would be a good idea to do some cleanup of `_tcpClient` before forgetting it, to ensure it is closed and can be garbage collected.